---
title: "01_wellness_profiling"
output: html_document
autor: "María Bueno Álvez"
date: "2024-12-13"
editor_options: 
  chunk_output_type: console
---

# Set up

```{r}
source("scripts/functions/functions_analyses.R")
source("scripts/functions/functions_utility.R")
source("scripts/functions/functions_visualization.R")
source("scripts/functions/themes_palettes.R")

# Read in data
data_wellness <- read_tsv("../Human-disease-blood-atlas/data/final_data/HPA/v24_2/wellness_data_ht_phase2.tsv")
meta_wellness <- read_tsv("../Human-disease-blood-atlas/data/final_data/HPA/v24_2/wellness_meta_ht_phase2.tsv")
concentrations <- import_df("data/other/concentrations_blood-atlas-build.xlsx")
olink_targets <- read_tsv("data/other/targets_olink_platforms.tsv")
secretome_hpa <- read_tsv("data/secretome_v24.tsv")
ht_blocks <- read_tsv("data/other/ht_assay_block.tsv")
```

# Wellness overview

```{r}
order_wellness <- 
  meta_wellness |> 
  filter(Visit == 1) |> 
  arrange(Date)

visit_ranges <- 
  meta_wellness %>%
  mutate(Visit = as.factor(Visit)) |> 
  group_by(Visit) %>%
  summarize(start_date = min(Date), end_date = max(Date))

plot_wellness <- 
  meta_wellness |> 
  mutate(Subject = factor(Subject, levels = order_wellness$Subject),
         Visit = as.factor(Visit)) |>
  ggplot(aes(Date, Subject, color = Visit)) +
  geom_rect(data = visit_ranges,
            aes(xmin = start_date, xmax = end_date, ymin = -Inf, ymax = Inf, fill = Visit),
            alpha = 0.3, inherit.aes = FALSE) +
  geom_line(show.legend = F, color = "grey80") +
  geom_point(show.legend = F) +
  # scale_color_manual(values = pal_wellness_individuals) +
  scale_fill_viridis_d(option = "viridis", direction = -1) +
  scale_color_viridis_d(option = "viridis", direction = -1) +
  theme_hpa(axis_y = T) +
  xlab("Sample date") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

plot_wellness
#ggsave(savepath("wellness_visit_timeline.pdf"), width = 9, height = 5)
```

# LOD

Analyze the number of proteins consistently detected across all visits for each individual.
High consistency in detection rates across visits and individuals supports technical reliability and biological stability.

## Per sample

```{r}
lod_sample_dat <- 
  data_wellness |> 
  mutate(above_LOD = ifelse(NPX > LOD, "Yes", "No")) |> 
  group_by(DAid) |> 
  count(above_LOD) 

lod_sample_dat_order <- 
  lod_sample_dat |> 
  filter(above_LOD == "Yes") |>
  arrange(-n)

lod_sample_dat|> 
  mutate(DAid = factor(DAid, levels = lod_sample_dat_order$DAid)) |>
  ggplot(aes(DAid, n, fill = above_LOD)) +
  geom_col() +
  theme_hpa(axis_x = F)

#ggsave(savepath("LOD_per_sample.png"), h = 5, w = 15)
```


## Per subject 

```{r}
dat <- 
  data_wellness |> 
  mutate(above_LOD = ifelse(NPX > LOD, "Yes", "No")) |>
  left_join(meta_wellness |> 
              select(DAid, Subject, Visit), by = "DAid") |> 
  group_by(Subject, Assay) |> 
  summarize(above_LOD_count = sum(NPX > LOD), .groups = "drop") |> 
  ungroup() 

number_of_visits <- 
  meta_wellness |> 
  group_by(Subject) |> 
  summarise(Number_of_visits = n_distinct(Visit)) |> 
  arrange(Number_of_visits)  

lod_subject_dat <- 
  dat |> 
  left_join(number_of_visits, by = "Subject") |> 
  mutate(perc = above_LOD_count / Number_of_visits,
         class = case_when(perc == 0 ~ "None",
                           perc == 1 ~ "All",
                           perc > 0 & perc <= 0.5 ~ "Some: < 50%",
                           perc < 1 & perc > 0.5 ~ "Some: > 50%"))  |> 
  group_by(Subject) |> 
  count(class) |> 
  mutate(class = factor(class, levels = names(pal_detectability)))

average_all <- 
  lod_subject_dat |> 
  filter(class == "All") |> 
  pull(n) |> 
  mean() 

(average_all / length(unique(data_wellness$Assay))) * 100

average_none <- 
  lod_subject_dat |> 
  filter(class == "None") |> 
  pull(n) |> 
  mean()

(average_none / length(unique(data_wellness$Assay))) * 100

subject_order <- 
  lod_subject_dat |> 
  filter(class == "All") |> 
  arrange(-n)

wellness_ids <- 
  subject_order |> 
  distinct(Subject) |> 
  pull()

names(pal_wellness_individuals) <- wellness_ids

plot_lod <- 
  lod_subject_dat |> 
  mutate(Subject = factor(Subject, levels = subject_order$Subject)) |> 
  ggplot(aes(Subject, n, fill = class)) +
  geom_col() +
  scale_fill_manual(values = pal_detectability) +
  theme_hpa(angled = T, axis_x = F) + 
  xlab("Subject")

plot_id <- 
  subject_order |> 
  ungroup() |> 
  mutate(Subject = factor(Subject, levels = subject_order$Subject)) |>
  mutate(rank = rank(-n, ties.method = "first"),
         y = 1) |> 
  ggplot(aes(Subject, y, fill = Subject)) +
  geom_tile(color = "white", show.legend = F) +
  scale_fill_manual(values = pal_wellness_individuals) +
  theme_hpa(axis_y = F) +
  coord_fixed() +
  theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank())

plot_lod / plot_id +
  plot_layout(heights = c(10,1))
ggsave(savepath_results("Publication-figures", "LOD_per_sample_names.pdf"), h = 4, w = 15) 
```

## Save detectability

```{r}
protein_detectability <- 
  data_wellness |>
  mutate(above_LOD = ifelse(NPX > LOD, "Yes", "No")) |>
  group_by(Assay, above_LOD) |> 
  summarise(count = n(), .groups = "drop") |> 
  complete(Assay, above_LOD = c("Yes", "No"), fill = list(count = 0)) |> 
  filter(above_LOD == "Yes") |> 
  mutate(`Percentage of samples above LOD` = (count/length(unique(meta_wellness$DAid)))*100) |> 
  select(-above_LOD) |> 
  rename(`Number of samples above LOD`= count) |> 
  arrange(-`Number of samples above LOD`)  

# protein_detectability |> 
#   left_join(olink_targets, by = "Assay") |> 
#   mutate(Platform = case_when(Platform == "HT" ~ "Olink Explore HT",
#                               Platform == "3K" ~ "Olink Explore 3072",
#                               Platform == "1.5K" ~ "Olink Explore 1463")) |> 
#   ggplot(aes(`Percentage of samples above LOD`, fill = Platform)) + 
#   geom_histogram() +
#   scale_fill_manual(values = pal_platforms) +
#   theme_hpa()
# 
# ggsave(savepath_results("Manuscript-figures", "S1.pdf"), h = 5, w = 7)

write_csv2(protein_detectability, savepath_data("Supplementary-files", "DataS1.csv"))
```

# Secretome analyses

```{r}
proteins_above_LOD_80 <- 
  data_wellness |>
  mutate(above_LOD = ifelse(NPX > LOD, "Yes", "No")) |>
  group_by(Assay) |> 
  count(above_LOD)  |> 
  arrange(-n) |> 
  mutate(n = (n/nrow(meta_wellness))*100) |> 
  filter(above_LOD == "Yes", 
         n >80)

(nrow(proteins_above_LOD_80) / length(unique(data_wellness$Assay))) * 100

proteins_below_LOD_80 <- 
  data_wellness |> 
  mutate(above_LOD = ifelse(NPX > LOD, "Yes", "No")) |>
  group_by(Assay) |> 
  count(above_LOD)  |> 
  arrange(-n) |> 
  mutate(n = (n/nrow(meta_wellness))*100) |> 
  filter(above_LOD == "No", 
         n >80)

(nrow(proteins_below_LOD_80) / length(unique(data_wellness$Assay))) * 100


secretome_donut <- 
  plot_donut(proteins_above_LOD_80$Assay, type = "secretome", legend = F) /
  plot_donut(proteins_below_LOD_80$Assay, type = "secretome") 

platform_donut <- 
  plot_donut(proteins_above_LOD_80$Assay, type = "platform", legend = F) /
  plot_donut(proteins_below_LOD_80$Assay, type = "platform") 

block_donut <- 
  plot_donut(proteins_above_LOD_80$Assay, type = "block", legend = F) /
  plot_donut(proteins_below_LOD_80$Assay, type = "block") 


(secretome_donut | platform_donut | block_donut) +
  plot_layout(guides = 'collect')
ggsave(savepath_results("Manuscript-figures", "FigS1b.pdf"), h = 10, w = 10)
```

## Histogram

```{r}
dat <- 
  secretome_hpa |> 
  filter(Gene %in% unique(data_wellness$Assay)) |> 
  select(Gene, `Secretome location`) |> 
  left_join(protein_detectability, by = c("Gene" = "Assay")) |> 
  mutate(`Secretome location` = ifelse(is.na(`Secretome location`), "Not secreted", `Secretome location`)) |> 
  filter(`Secretome location` != "Immunoglobulin genes") 

dat |> 
  distinct(`Secretome location`)
  
secretome_level <- c("Not secreted",
                    "Secreted to blood",
                    "Intracellular and membrane",
                    "Secreted to extracellular matrix",
                    "Secreted in brain",
                    "Secreted to digestive system",
                    "Secreted in female reproductive system",
                    "Secreted in male reproductive system",
                    "Secreted in other tissues",
                    "Secreted - unknown location")

dat |> 
  mutate(`Secretome location` = factor(`Secretome location`, levels = secretome_level)) |>
ggplot(aes(`Percentage of samples above LOD`, fill = `Secretome location`)) + 
  geom_histogram() +
  scale_fill_manual(values = pal_secreted) +
  facet_wrap(~`Secretome location`, scales = "free", nrow = 2) +
  theme_hpa() +
  theme(legend.position = "top")


ggsave(savepath_results("Manuscript-figures", "S1a.pdf"), h = 7, w = 12)

```

## Secretome classes

```{r}
secreted_other <- c("Secreted in other tissues",
                    "Secreted - unknown location",
                    "Secreted in male reproductive system",
                    "Secreted in female reproductive system")

data_wellness |>
  mutate(above_LOD = ifelse(NPX > LOD, "Yes", "No")) |> 
  left_join(secretome_hpa |> 
  filter(Gene %in% unique(data_wellness$Assay)) |> 
  select(Assay = Gene, `Secretome location`)) |> 
  filter(`Secretome location` != "Immunoglobulin genes") |>
  mutate(`Secretome location` = ifelse(is.na(`Secretome location`), "Not secreted", `Secretome location`),
         `Secretome location` = ifelse(`Secretome location` == "Intracellular and membrane", "Not secreted", `Secretome location`),
         `Secretome location` = ifelse (`Secretome location` %in% secreted_other, "Secreted to other locations", `Secretome location`)) |> 
  group_by(`Secretome location`, Assay) |> 
  count(above_LOD) |> 
  mutate(percent = n / sum(n) * 100) |>   # Convert n to percentage
  ungroup() |> 
  group_by(`Secretome location`, Assay) |> 
  mutate(percent_yes = sum(percent[above_LOD == "Yes"])) |>  # Calculate % of "Yes"
  filter(`Secretome location` %in% c("Secreted to blood", "Not secreted")) |> 
  ggplot(aes(x = fct_reorder(Assay, -percent_yes), y = percent, fill = above_LOD)) +  
  geom_col() +
  facet_wrap(~`Secretome location`, scales = "free_x", ncol = 1) +
  theme_hpa(axis_x = F) +
  ylab("Percentage of samples")

```

## Histogram

```{r}
protein_detectability |> 
  left_join(secretome_hpa, by = c("Assay" = "Gene")) |> 
  mutate(`Secretome location` = case_when(is.na(`Secretome location`) ~ "Not secreted", 
                                          `Secretome location` == "Intracellular and membrane" ~ "Not secreted", 
                                          `Secretome location`== "Secreted to blood" ~ "Actively secreted to blood",
                                          T ~ "Secreted to other locations")) |> 
  mutate(`Secretome location` = factor(`Secretome location`, levels = c("Actively secreted to blood", "Secreted to other locations", "Not secreted"))) |> 
  ggplot(aes(`Percentage of samples above LOD`, fill = `Secretome location`)) + 
  geom_histogram() +
  scale_fill_manual(values = pal_secretome_condensed) +
  theme_hpa()

#ggsave(savepath_results("Manuscript-figures", "secretome_LOD.pdf"), h = 6, w = 7)
```


# Remove proteins below LOD in > 80% of samples

```{r}
data_wellness_detected <- 
  data_wellness |>
  filter(!Assay %in% proteins_below_LOD_80$Assay)
```

# UMAP

Visualize stability of the proteome per sample.

```{r}
# wellness_umap <- 
#   do_umap(data = data_wellness_detected, 
#           wide = F, 
#           impute = T, 
#           plots = F)

wellness_umap_all_proteins <- 
  do_umap(data = data_wellness, 
          wide = F, 
          impute = T, 
          plots = F)

wellness_umap_all_proteins |> 
  left_join(meta_wellness |> 
              select(DAid, Subject, Visit), by =c("Sample" = "DAid")) |>
  ggplot(aes(UMAP1, UMAP2, color = Subject, frame = Visit)) +
  geom_point(show.legend = F) +
  
  scale_color_manual(values = pal_wellness_individuals) +
  theme_hpa()
ggsave(savepath_results("Manuscript-figures", "Fig1f_all_proteins.pdf"), h = 6, w = 6)

wellness_umap_awellness_umapll_proteins + 
  facet_wrap(~Visit)

ggsave(savepath_results("Manuscript-figures", "Fig1f_all_proteins_faceted.pdf"), h = 8, w = 12)
```

## Interactive visualization

```{r}
dynamic_plot_data <- 
  wellness_umap |> 
  rename(DAid = Sample) |> 
  left_join(wellness_meta, by = "DAid") 

# Create the ggplot object
p <- ggplot(dynamic_plot_data, aes(x = UMAP1, y = UMAP2, color = subject, frame = visit)) +
  geom_point(show.legend = F) +
  theme_hpa() +
  labs(x = "UMAP1", y = "UMAP2")

# Convert ggplot to a plotly object for interactivity
p_dynamic <- ggplotly(p)

# Display the dynamic plot
p_dynamic
```


# Mixed models

Identify proteins that are significantly differentially expressed between visits. Summarize as volcano plot & boxplots. 

Reinforce stability of the proteome by showing that very few proteins change over time. Highlight how few proteins vary across visits.

```{r}
# Prepare data for mixed effect modelling
data_mixed_model <- 
  data_wellness_detected |> 
  select(DAid, Assay, NPX) %>%
  left_join(meta_wellness |> 
              mutate(Visit = as.numeric(Visit)) |> 
              select(DAid, Subject, Age, Sex), by = "DAid")

# Run mixed effect models on all proteins extracting the fixed effects
fixed_effects <- 
  map_df(unique(data_mixed_model$Assay), function(protein) {
    
    do_mixed_effect_model(df = data_mixed_model, 
                          type = "fixed_effects",
                          protein = protein)
    
  })

# Run mixed effect models on all proteins extracting the variance explained
variance_explained <- 
  map_df(unique(data_mixed_model$Assay), function(protein) {
    
    do_mixed_effect_model(df = data_mixed_model, 
                          type = "variance_explained",
                          protein = protein)
    
  })

saveRDS(fixed_effects, "data/processed/MM_estimates_Wellness.rds")
saveRDS(variance_explained, "data/processed/MM_variance_Wellness.rds")

write_excel_csv(fixed_effects, savepath_data("Supplementary-files", "Data_S2_Wellness_estimates.csv"))
write_excel_csv(variance_explained, savepath_data("Supplementary-files", "Data_S2_Wellness_variance.csv"))
```

## Visualize 

### Overall

```{r}
order <- 
  variance_explained |> 
  filter(Component == "Residual") |> 
  arrange(Variance)

top_proteins <- 
  variance_explained |> 
  filter(Component == "Random effects (subject)") |> 
  arrange(-Variance) |> 
  head(5) |> 
  bind_rows(variance_explained |> 
              filter(Component == "Fixed effects (age & sex)") |> 
              arrange(-Variance) |> 
              head(5)) |> 
  mutate(Variance = Variance * 100) 

variance_explained |> 
  mutate(Protein = factor(Protein,
                          levels = order$Protein)) |> 
  filter(Component != "Residual") |> 
  mutate(Variance = Variance * 100) |> 
  ggplot(aes(Protein, Variance, fill = Component)) +
  geom_col(position = "stack") +
  scale_fill_manual(values = pal_mm) +
  theme_hpa() +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank()) +
  ylab("Variance explained (%)")

ggsave(savepath_results("Manuscript-figures", "MM-variance-explained.pdf"), h = 4, w = 14)
```

### Examples main

```{r}
# Example individual
data_wellness |> 
  filter(Assay %in% c("PNLIPRP2", "SPINT3")) |> 
  left_join(meta_wellness, by = "DAid") |> 
  mutate(Visit = factor(Visit)) |> 
  ggplot(aes(Visit, NPX, color = Sex, group = Subject)) +
  geom_point() +
  geom_line(show.legend = F, alpha = 7) +
  scale_color_manual(values = pal_sex) +
  facet_wrap(~Assay, nrow = 1, scale = "free") +
  theme_hpa() +
  theme(legend.position = "top")

ggsave(savepath_results("Manuscript-figures", "MM-example-individual.pdf"), h = 3.5, w = 6)
```


### Examples supplementary

```{r}
top_100 <- 
  order |> 
  mutate(Variance = Variance * 100) |> 
  head(100)

p1 <- 
  variance_explained |> 
  filter(Protein %in% top_100$Protein) |>
  mutate(Protein = factor(Protein, levels = top_100$Protein)) |> 
  filter(Component != "Residual") |> 
  mutate(Variance = Variance * 100) |> 
  ggplot(aes(Protein, Variance, fill = Component)) +
  geom_col(position = "stack") +
  scale_fill_manual(values = pal_mm) +
  theme_hpa(angled = T) +
  # theme(legend.position = "top",
  #       axis.ticks.x = element_blank(),
  #       axis.text.x = element_blank()) +
  ylab("Variance explained (%)") +
  # Add labels for top proteins outside the plot, spread on the x-axis
  geom_label_repel(data = top_proteins, aes(label = Protein, fill = Component),
             nudge_x = 0.25,  # Move labels to the right of the plot
             color = "black", fontface = "bold", size = 4, 
             hjust = 0, vjust = 0.5)  #

proteins_random <- 
  top_proteins |> 
  filter(Component == "Random effects (subject)") 

p2 <- 
  data_wellness |> 
  filter(Assay %in% proteins_random$Protein) |> 
  mutate(Assay = factor(Assay, levels = proteins_random$Protein)) |>
  left_join(meta_wellness, by = "DAid") |> 
  ggplot(aes(Visit, NPX, color = Subject)) +
  geom_point(show.legend = F) +
  geom_line(show.legend = F) +
  scale_color_manual(values = pal_wellness_individuals) +
  facet_wrap(~Assay, nrow = 1, scale = "free") +
  theme_hpa()

# proteins_fixed <- 
#   top_proteins |> 
#   filter(Component == "Fixed effects (age & sex)") 
# 
# fixed_effects |> 
#   filter(Assay %in% proteins_fixed$Protein)

top_age <- 
  fixed_effects |> 
  mutate(p.adjusted = p.adjust(p.value, method = "BH")) |> 
  filter(p.adjusted < 0.05) |> 
  filter(term == "Age") |> 
  arrange(p.adjusted)|> 
  head(5)

top_sex <- 
  fixed_effects |> 
  mutate(p.adjusted = p.adjust(p.value, method = "BH")) |> 
  filter(p.adjusted < 0.05) |> 
  filter(term == "SexM") |> 
  arrange(p.adjusted) |> 
  head(5)

p3 <- 
  data_wellness |> 
  filter(Assay %in% top_sex$Assay) |> 
  mutate(Assay = factor(Assay, levels = top_sex$Assay)) |>
  left_join(meta_wellness, by = "DAid") |> 
  ggplot(aes(Visit, NPX, color = Sex, group = Subject)) +
  geom_point(show.legend = F) +
  geom_line(show.legend = F) +
  scale_color_manual(values = pal_sex) +
  facet_wrap(~Assay, nrow = 1, scale = "free") +
  theme_hpa()

p4 <- 
  data_wellness |> 
  filter(Assay %in% top_age$Assay) |> 
  mutate(Assay = factor(Assay, levels = top_age$Assay)) |>
  left_join(meta_wellness, by = "DAid") |> 
  ggplot(aes(Age, NPX, color = Age)) +
  geom_point() +
  geom_smooth(color = "black") +
  scale_color_viridis_c(option = "A") +
  facet_wrap(~Assay, nrow = 1, scale = "free") +
  theme_hpa()

p1 / p2 / p3 / p4
ggsave(savepath_results("Manuscript-figures", "MM-variance-explained.pdf"), h = 12, w = 14)
```


### Example age

```{r}
dat_age <- 
  mm_all_proteins |> 
  filter(term == "age") |> 
  arrange(p.value) |> 
  mutate(sig = case_when(p.value< 0.06 & estimate > 0 ~ "significant up",
                         p.value< 0.06 & estimate < 0 ~ "significant down",
                         TRUE ~ "not significant"))
labels_age <- 
  dat_age |> 
  top_n(n = 10, wt = -log10(p.value)) 

volcano_plot_age <- 
  dat_age |> 
  ggplot(aes(x = estimate, y = -log10(p.value), color = sig, label = Protein)) +
  geom_point(size = 1, alpha = 0.4, show.legend = F) + 
  geom_text_repel(data = labels_age, size = 2, show.legend = F) +
  geom_hline(yintercept = -log10(0.05), linetype = 'dashed', color = "darkgrey") +
  #geom_vline(xintercept = -1, linetype = 'dashed', color = "darkgrey") +
 # geom_vline(xintercept = 1, linetype = 'dashed', color = "darkgrey") +
  scale_color_manual(values = pal_de) +
  theme_hpa() +   
  theme(axis.text = element_text(size = 8),
        legend.position = "top") 
```

### Sex

```{r}
dat_sex <- 
  mm_all_proteins |> 
  filter(term == "sexM") |> 
  arrange(p.value) |> 
  mutate(sig = case_when(p.value< 0.06 & estimate > 0 ~ "significant up",
                         p.value< 0.06 & estimate < 0 ~ "significant down",
                         TRUE ~ "not significant"))
labels_sex <- 
  dat_sex |> 
  top_n(n = 10, wt = -log10(p.value)) 

volcano_plot_sex <- 
  dat_sex |> 
  ggplot(aes(x = estimate, y = -log10(p.value), color = sig, label = Protein)) +
  geom_point(size = 1, alpha = 0.4, show.legend = F) + 
  geom_text_repel(data = labels_sex, size = 2, show.legend = F) +
  geom_hline(yintercept = -log10(0.05), linetype = 'dashed', color = "darkgrey") +
  #geom_vline(xintercept = -1, linetype = 'dashed', color = "darkgrey") +
 # geom_vline(xintercept = 1, linetype = 'dashed', color = "darkgrey") +
  scale_color_manual(values = pal_de) +
  theme_hpa() +   
  theme(axis.text = element_text(size = 8),
        legend.position = "top") 

volcano_plot_sex + volcano_plot_age
ggsave(savepath("volcanos_wellness_MM.pdf"), h = 4, w = 10)

volcano_plot_age
ggsave(savepath("volcano_age_MM_wellness.pdf"), h = 4, w = 5)
```

### Proteins age

```{r}
top_age_proteins <- 
  dat_age |> 
  group_by(sig) |>
  top_n(n = 4, wt = -log10(p.value)) |> 
  filter(sig != "not significant") 

wellness_data |> 
  filter(Assay %in% top_age_proteins$Protein) |>
  left_join(wellness_meta |> 
              select(DAid, subject, age, visit)) |> 
  mutate(Assay = factor(Assay, levels = top_age_proteins$Protein)) |>
  ggplot(aes(age, NPX, color = age)) +
  geom_line(aes(group = subject), alpha = 0.3, color = "grey80") +
  geom_point() +
  geom_smooth(method = "lm", se = F, color = "white") +
  #geom_quasirandom(alpha = 0.7, show.legend = F) +
 # geom_boxplot(fill = NA, outlier.colour = NA, color = "grey20") +
 # scale_color_manual(values = pal_bamse) +
  scale_color_viridis_c(option = "magma") +
  facet_wrap(~Assay, scales = "free_y", nrow = 1) +
  theme_hpa() 
  
ggsave(savepath("top_age_proteins_bamse.pdf"), h = 2.5, w = 12)

wellness_data |> 
  filter(Assay %in% top_age_proteins$Protein) |>
  left_join(wellness_meta |> 
              select(DAid, subject, age, visit)) |> 
  mutate(Assay = factor(Assay, levels = top_age_proteins$Protein)) |>
  ggplot(aes(visit, NPX, color = visit)) +
  geom_line(aes(group = subject), alpha = 0.3, color = "grey80") +
  geom_point() +
  #geom_smooth(method = "lm", se = F, color = "white") +
  geom_quasirandom(alpha = 0.7, show.legend = F) +
  geom_boxplot(fill = NA, outlier.colour = NA, color = "grey20") +
 # scale_color_manual(values = pal_bamse) +
  scale_color_viridis_d(option = "magma") +
  facet_wrap(~Assay, scales = "free_y", nrow = 2) +
  theme_hpa(axis_x = F) 
  
ggsave(savepath("top_age_proteins_wellness_d.pdf"), h = 2.5, w = 12)
ggsave(savepath("top_age_proteins_wellness_d.pdf"), h = 3.5, w = 8)


top_sex_proteins <- 
  dat_sex |> 
  group_by(sig) |>
  top_n(n = 4, wt = -log10(p.value)) |> 
  filter(sig != "not significant") 

bamse_data |> 
  filter(Assay %in% top_sex_proteins$Protein) |>
  left_join(bamse_meta |> 
              select(DAid, subject, Sex)) |> 
  mutate(Assay = factor(Assay, levels = top_sex_proteins$Protein)) |>
  ggplot(aes(Sex, NPX, color = Sex)) +
  geom_line(aes(group = subject), alpha = 0.3, color = "grey80") +
  geom_quasirandom(alpha = 0.7, show.legend = F) +
  geom_boxplot(fill = NA, outlier.colour = NA, color = "grey20") +
  #scale_color_manual(values = pal_bamse) +
  facet_wrap(~Assay, scales = "free_y", nrow = 1) +
  theme_hpa(axis_x = F) 

ggsave(savepath("top_sex_proteins_wellness.pdf"), h = 2.5, w = 12)
```